lightning/src/routing/router.rs

```diff
@@ -1512,6 +1515,66 @@ where L::Target: Logger {
 		}
 	}
 
+	// Step (10).
+	// Add a random 'shadow route' offset to the CLTV expiry delta values to improve privacy.
+	for path in selected_paths.iter_mut() {
+		let mut shadow_ctlv_expiry_delta_offset: u32 = 0;
+
+		// Choose the last publicly known node as the starting point for the random walk
+		let mut cur_node_iter = path.iter().rev();
+		let mut cur_node_id = NodeId::from_pubkey(&cur_node_iter.next().unwrap().as_ref().unwrap().pubkey);
+		assert_eq!(cur_node_id, payee_node_id);
+		while !network_nodes.contains_key(&cur_node_id) {
+			cur_node_id = NodeId::from_pubkey(&cur_node_iter.next().unwrap().as_ref().unwrap().pubkey);
```

Comment:
Right, this panics if the full route is private, right?

Reply:
Yep, this is one reason why I deemed this approach 'insufficient'. Fixed in https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/48c7e95ee1a7861681ea93f13aca42f2d835afdf.